### PR TITLE
Feat phone auth 360dialog

### PR DESCRIPTION
**Pull request by aakashcodess**
## What does this PR do?

This PR will add phone authentication with whatsapp using 360dialog.

## Test Plan

To test you have to create a phoneSession.
To test you should have a live 360dialog api or it can be tested with [360dialog sandbox api](https://docs.360dialog.com/docs/whatsapp-api/whatsapp-api/sandbox) 
In .env file
_APP_SMS_PROVIDER = sms://template_name:api_key@dialog360
_APP_PHONE_FROM = template_namespace

## Related PRs and Issues

NO

### Have you added your change to the [Changelog](https://github.com/appwrite/appwrite/blob/master/CHANGES.md)?

N0

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes

### CODE REVIEW

**Comment by aakashcodess** [comment]
360dialog takes 4 parameters to send a message Api-key, template name, template namespace, template language. 
for now i have hard coded language to english on [L41](https://github.com/appwrite/appwrite/blob/b949b7785fdf0b6bab05216602154862a19fc5b8/src/Appwrite/SMS/Adapter/Dialog360.php#L41 )

More details regarding templates can be found on https://developers.facebook.com/docs/whatsapp/api/messages/message-templates/

**Comment by aakashcodess** [comment]
We can get template name, template namespace, template language for whatsapp business template settings. Added screenshots from whatsapp/facebook business manager templates settings page. 
But first we have to apply for a whatsapp template with one variable in our template which will get replaced with secret from appwrite when sending message. (added screenshot of both approved template and received message)
whatsapp-template-example
Screenshot 2022-10-20 at 12 17 59 AM
Screenshot 2022-10-20 at 12 18 21 AM

File: src/Appwrite/SMS/Adapter/Dialog360.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+// Reference Material
+// https://docs.360dialog.com/docs/whatsapp-api/whatsapp-api/sandbox#php
+
+class Dialog360 extends Adapter
+{
+    /**
+     * @var string
+     */
+    private string $endpoint = 'https://waba.360dialog.io/v1/messages';
```
Comment (stnguyen90) [inline src/Appwrite/SMS/Adapter/Dialog360.php:15]: Can you actually make 2 Adapters: 1 called Dialog360 and the other called Dialog360Sandbox. The difference would be the endpoint? I think it makes sense for 1 to be a subclass of the other.

File: src/Appwrite/SMS/Adapter/Dialog360.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
+<?php
+
+namespace Appwrite\SMS\Adapter;
+
+use Appwrite\SMS\Adapter;
+
+// Reference Material
+// https://docs.360dialog.com/docs/whatsapp-api/whatsapp-api/sandbox#php
+
+class Dialog360 extends Adapter
```
Comment (stnguyen90) [inline src/Appwrite/SMS/Adapter/Dialog360.php:10]: Would you please implement this in [utopia-php/messaging](https://github.com/utopia-php/messaging) following [these](https://github.com/utopia-php/messaging/blob/main/docs/add-new-adapter.md) instructions?

File: src/Appwrite/SMS/Adapter/Dialog360.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    public function send(string $from, string $to, string $message): void
+    {
+        $to = ltrim($to, '+');
+
+        $this->request(
+            method: 'POST',
+            url: $this->endpoint,
+            headers: ["Content-Type: application/json",
+                      "D360-Api-Key: {$this->secret}"],
```
Comment (stnguyen90) [inline src/Appwrite/SMS/Adapter/Dialog360.php:31]: Some more style changes:

```suggestion
            headers: [
                "Content-Type: application/json",
                "D360-Api-Key: {$this->secret}"
            ],
```

File: src/Appwrite/SMS/Adapter/Dialog360.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                                [
+                                    'type' => 'text',
+                                    'text' => $message
+                                ]
+                            ]
+                        ]
+                    ]
+                ]
+                ]
+            )
```
Comment (stnguyen90) [inline src/Appwrite/SMS/Adapter/Dialog360.php:56]: Let's fix this:

```suggestion
            ])
```

**Review by stnguyen90 — CHANGES_REQUESTED** [review]
This is AMAZING! 🤯 I've left some request for changes. Please make sure to create a PR in the Utopia Messaging library. We can leave this for later when we update Appwrite to use the Messaging library.

File: src/Appwrite/SMS/Adapter/Dialog360.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            payload: json_encode([
+                'recipient_type' => 'individual',
+                'to' => $to,
+                'type' => 'template',
+                'template' => [
+                    'namespace' => $from,
+                    'name' => "{$this->user}",
+                    'language' => [
+                        'policy' => 'deterministic',
+                        'code' => 'en'
```
Comment (stnguyen90) [inline src/Appwrite/SMS/Adapter/Dialog360.php:41]: In the messaging library, language can be another parameter to the constructor of the adapter.

**Comment by aakashcodess** [comment]
Hey @stnguyen90, I have made the requested changes and also added the sandbox adapter. Will add the adapter in Utopia Messaging library too.

File: src/Appwrite/SMS/Adapter/Dialog360.php (hunk 5 above)
Comment (aakashcodess) [reply src/Appwrite/SMS/Adapter/Dialog360.php:43]: Yes, I will do it. Thanks for the suggestion.

**Review by stnguyen90 — COMMENTED** [review]
> I have made the requested changes and also added the sandbox adapter. Will add the adapter in Utopia Messaging library too.

@aakashcodess, sorry I wasn't suggesting to make edits in this PR. It might be best to close this PR and just add the new Adapters in the Utopia Messaging library.

Btw, when making your chan get in the Utopia Messaging library, make sure to have 1 Adapter subclass the other to reduce the duplicate code.

**Comment by aakashcodess** [comment]
@stnguyen90 Thanks for the update. I will add new adapters in messaging library.

**Comment by stnguyen90** [comment]
Adding `hacktoberfest-accepted` so you get some credit for Hacktoberfest. We'll clean this up and merge it later.

**Comment by christyjacob4** [comment]
@aakashcodess thanks a lot for your contributions during Hacktoberfest 2022!  

Please reach out to me on our [Discord server](https://appwrite.io/discord) if you would like to claim your Appwrite swags! As a way of saying thank you, we would also love to invite you to join the Appwrite organization on GitHub. Please share your GitHub username with us on Discord. 

Also let us know if you're willing to continue working on it?

**Comment by aakashcodess** [comment]
@christyjacob4 yes, I will resume working on it this weekend. I will message you on discord to join the github org.

**Comment by gewenyu99** [comment]
@aakashcodess How's it going with this auth provider? Can we do anything to help?

**Comment by aakashcodess** [comment]
> @aakashcodess How's it going with this auth provider? Can we do anything to help?

Hey @gewenyu99 , I was on break, I will start working on it. If I face any issue, I will ping you on discord.

**Comment by aakashcodess** [comment]
Hey @stnguyen90 @gewenyu99 , I wanted to finish this. I didn't had access to live 360dialog api, now i do have access to 360dialog api key, can I continue to work on this. Or i can add the adapter (twilio whatsapp and 360dialog whatsapp) in messaging library.

**Pull request closed without merging**